### PR TITLE
link: fix false positive crtbegin/crtend detection

### PR DESCRIPTION
**Pull request by andrewrk**
Embrace the Path abstraction, doing more operations based on directory handles rather than absolute file paths. Most of the diff noise here comes from this one.

Fix sorting of crtbegin/crtend atoms. Previously it would look at all path components for those strings.

Make the C runtime path detection partially a pure function, and move some logic to glibc.zig where it belongs.

### CODE REVIEW

**Comment by andrewrk** [comment]
```
test-link
└─ link_test_cases
   └─ test-elf
      └─ test-elf-emit-static-lib-x86_64-linux.4.19...6.10.3-musl-Debug-llvm-no-lld
         └─ CheckObject failure
error: 
========= expected to find: ==========================
in object /home/andy/src/zig/.zig-cache/o/fc09b17a5ef82aeafeea48b5d4cf5938/obj1.o
========= but parsed file does not contain it: =======
archive symbol table
in object obj1.o
fooBar
foo
bar
in object obj2.o
tentative
in object a_very_long_file_name_so_that_it_ends_up_in_strtab.o
weakFoo
strongBar
strongBarAlias

======================================================
```
I'm not sure what to make of this output, and the test failure does not set up a reproduction. So, the next step here will be improving the test harness along with solving #21637.

File: src/link/Elf/file.zig [hunk 1]
```diff
@@ -240,38 +240,39 @@ pub const File = union(enum) {
...
+            .zig_object => |zo| {
+                const basename = zo.basename;
+                if (basename.len <= Archive.max_member_name_len) return;
+                zo.output_ar_state.name_off = try ar_strtab.insert(allocator, basename);
+            },
+            .object => |o| {
+                const basename = std.fs.path.basename(o.path.sub_path);
+                if (basename.len <= Archive.max_member_name_len) return;
+                o.output_ar_state.name_off = try ar_strtab.insert(allocator, basename);
+            },
```
Comment (andrewrk) [inline src/link/Elf/file.zig:258]: @kubkon archive files cannot have path components in them, so I do basename here. I believe that is causing the test to fail.

Comment (kubkon) [reply src/link/Elf/file.zig:258]: I think you are absolutely right.

**Comment by andrewrk** [comment]
image

Oops! :scream:

**Comment by kubkon** [comment]
> image
> 
> Oops! 😱

Wait, is this not a fix to https://github.com/ziglang/zig/issues/21637 by chance?

**Comment by kubkon** [comment]
Ahh, sadly it is not, that would be too amazing 😁

**Comment by andrewrk** [comment]
No no, just a mistake I made locally yesterday.

**Pull request merged**